src/utils/post_list.js

```diff
@@ -0,0 +1,404 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import {createSelector} from 'reselect';
+
+import {Posts, Preferences} from 'constants';
+
+import {makeGetPostsForIds} from 'selectors/entities/posts';
+import {getBool} from 'selectors/entities/preferences';
+import {getCurrentUser} from 'selectors/entities/users';
+
+import {createIdsSelector, memoizeResult} from 'utils/helpers';
+import {isUserActivityPost, shouldFilterJoinLeavePost} from 'utils/post_utils';
+
+export const COMBINED_USER_ACTIVITY = 'user-activity-';
+export const DATE_LINE = 'date-';
+export const START_OF_NEW_MESSAGES = 'start-of-new-messages';
+
+const MAX_COMBINED_SYSTEM_POSTS = 100;
+
+function shouldShowJoinLeaveMessages(state) {
+    // This setting is true or not set if join/leave messages are to be displayed
+    return getBool(state, Preferences.CATEGORY_ADVANCED_SETTINGS, Preferences.ADVANCED_FILTER_JOIN_LEAVE, true);
+}
+
+export function makePreparePostIdsForPostList() {
+    const filterPostsAndAddSeparators = makeFilterPostsAndAddSeparators();
+    const combineUserActivityPosts = makeCombineUserActivityPosts();
+
+    return (state, options) => {
+        let postIds = filterPostsAndAddSeparators(state, options);
+        postIds = combineUserActivityPosts(state, postIds);
+
+        return postIds;
+    };
+}
+
+// Returns a selector that, given the state and an object containing an array of postIds and an optional
+// timestamp of when the channel was last read, returns a memoized array of postIds interspersed with
+// day indicators and an optional new message indicator.
+export function makeFilterPostsAndAddSeparators() {
+    const getPostsForIds = makeGetPostsForIds();
+
+    return createIdsSelector(
+        (state, {postIds}) => getPostsForIds(state, postIds),
+        (state, {lastViewedAt}) => lastViewedAt,
+        (state, {indicateNewMessages}) => indicateNewMessages,
+        (state) => state.entities.posts.selectedPostId,
+        getCurrentUser,
+        shouldShowJoinLeaveMessages,
+        (posts, lastViewedAt, indicateNewMessages, selectedPostId, currentUser, showJoinLeave) => {
+            if (posts.length === 0 || !currentUser) {
+                return [];
```

Comment:
How about returning `null`?

Reply:
This matches the current version of the code [here](https://github.com/mattermost/mattermost-mobile/blob/master/app/selectors/post_list.js#L35), but I'm 0/5 on changing it or leaving it the same